x509-cert/src/serial_number.rs

```diff
@@ -65,6 +79,104 @@ impl<P: Profile> SerialNumber<P> {
+    /// of output from the CSPRNG.
+    ///
+    /// [ballot 164]: https://cabforum.org/2016/03/31/ballot-164/
+    #[cfg(feature = "builder")]
+    pub fn generate_with_length<N>(rng: &mut impl CryptoRngCore) -> Result<Self>
+    where
+        N: Unsigned + ArrayLength<u8>,
+        // Rand minimum is 64 bits
+        N: Min<U8>,
+        // Max length is 20 bytes, encoding of sign bit might
+        // requires an additional byte, so the limit is 19.
+        N: Max<U19>,
+    {
+        Self::generate_with_prefix::<UTerm, N>(GenericArray::default(), rng)
+    }
+
+    /// Generates a random serial number from RNG.
+    ///
+    /// This follows the recommendation the CAB forum [ballot 164] and uses a minimum of 64 bits
+    /// of output from the CSPRNG.
+    ///
+    /// [ballot 164]: https://cabforum.org/2016/03/31/ballot-164/
+    #[cfg(feature = "builder")]
+    #[allow(unused_qualifications)]
+    pub fn generate_with_dyn_length(rng: &mut impl CryptoRngCore, len: usize) -> Result<Self> {
+        macro_rules! impl_generate {
+            ($len:literal => $u:ty) => {
+                if (len == $len) {
+                    return Self::generate_with_prefix::<UTerm, $u>(GenericArray::default(), rng);
+                }
+            };
+            ($len:literal => $u:ty, $($rest:tt)*) => {
+                impl_generate!($len => $u);
+                impl_generate!($($rest)*);
+            };
+        }
+
+        impl_generate!(
+            8 => consts::U8,
+            9 => consts::U9,
+            10 => consts::U10,
+            11 => consts::U11,
+            12 => consts::U12,
+            13 => consts::U13,
+            14 => consts::U14,
+            15 => consts::U15,
+            16 => consts::U16,
+            17 => consts::U17,
+            18 => consts::U18,
+            19 => consts::U19
+        );
+
+        Err(ErrorKind::Failed.into())
+    }
+
+    /// Generates a random serial number from RNG. Include a prefix value.
+    ///
+    /// This follows the recommendation the CAB forum [ballot 164] and uses a minimum of 64 bits
+    /// of output from the CSPRNG.
+    ///
+    /// [ballot 164]: https://cabforum.org/2016/03/31/ballot-164/
+    #[cfg(feature = "builder")]
+    pub fn generate_with_prefix<Prefix, N>(
+        prefix: GenericArray<u8, Prefix>,
+        rng: &mut impl CryptoRngCore,
+    ) -> Result<Self>
+    where
+        N: Unsigned,
+        Prefix: Unsigned + ArrayLength<u8>,
+        Prefix: Add<N>,
+        <Prefix as Add<N>>::Output: ArrayLength<u8>,
+        // Rand minimum is 64 bits
+        N: Min<U8>,
+        // Max length is 20 bytes, encoding of sign bit might
+        // requires an additional byte, so the limit is 19.
+        <Prefix as Add<N>>::Output: Max<U19>,
+    {
+        let mut buf = GenericArray::<_, <Prefix as Add<N>>::Output>::default();
+        buf[..Prefix::USIZE].copy_from_slice(&prefix);
+        rng.fill_bytes(&mut buf[Prefix::USIZE..]);
+
+        Self::new(&buf)
+    }
```

Comment:
If `generate` is the preferred API, is there really a good reason to have three different other functions for customizing the length/prefix?

Reply:
I, personally, intend to consume `generate_with_prefix`, but I'd happily use `generate` in tests cases too. Those two I'd like to keep.

I don't see a use for the non-const-generic api, but I don't claim to know all of the use-cases.

Reply:
I intend to use `generate_with_prefix` this one I'd like to keep.

You mentioned earlier the non const-generic API. I don't really have an opinion against it. I don't know :)
I could leave the implementation of `generate_with_dyn_length` in the rustdoc too.

`generate_with_length` could just go away too.

Reply:
Maybe get rid of the macro-based monomorphization and make the core API dynamic, and the const generic one written in terms of it? It just seems needlessly overcomplicated.